commcare_connect/users/user_credentials.py

```diff
@@ -0,0 +1,108 @@
+from django.db.models import Count, F, OuterRef, Q, Subquery
+
+from commcare_connect.opportunity.models import Assessment, CompletedWork, CompletedWorkStatus, CredentialConfiguration
+from commcare_connect.users.models import UserCredential
+
+
+class UserCredentialIssuer:
+    @classmethod
+    def run(cls):
+        all_cred_levels = UserCredential.DeliveryLevel.choices + UserCredential.LearnLevel.choices
+        user_credentials_data = []
+        for cred_level, _ in all_cred_levels:
+            cred_level_opportunity_ids = CredentialConfiguration.objects.filter(
+                Q(learn_level=cred_level) | Q(delivery_level=cred_level)
+            ).values_list("opportunity_id", flat=True)
+
+            if cred_level in UserCredential.DeliveryLevel.values:
+                user_credentials_data.extend(
+                    cls.get_delivery_user_credentials(cred_level_opportunity_ids, cred_level),
+                )
+            else:
+                user_credentials_data.extend(
+                    cls.get_learning_user_credentials(cred_level_opportunity_ids, cred_level),
+                )
+
+            UserCredential.objects.bulk_create(user_credentials_data, batch_size=100, ignore_conflicts=True)
+            user_credentials_data = []
+            # Todo: send to PersonalID for credential issuance
+            # Ticket: CCCT-1725
+
+    @classmethod
+    def get_delivery_user_credentials(cls, opportunities, credential_level):
+        level_int = UserCredential.delivery_level_num(credential_level, UserCredential.CredentialType.DELIVERY)
+        if not level_int:
+            return []
+
+        user_credentials_to_exclude = UserCredential.objects.filter(
+            opportunity=OuterRef("opportunity_access__opportunity"),
+            credential_type=UserCredential.CredentialType.DELIVERY,
+            level=credential_level,
+        ).values_list("user_id", flat=True)
```

Comment:
You can eliminate the need for this sub query to re-scan the User table by just returning the opportunity_access ids here and then using those opportunity access in the exclude below.